addons/sale/tests/test_reinvoice.py

```diff
@@ -278,3 +289,54 @@ def test_not_reinvoicing_invoiced_so_lines(self):
         self.assertFalse(so_line4, "No re-invoicing should have created a new sale line with product #2")
         self.assertEqual(so_line1.qty_delivered, 1, "No re-invoicing should have impacted exising SO line 1")
         self.assertEqual(so_line2.qty_delivered, 1, "No re-invoicing should have impacted exising SO line 2")
+
+    def test_refund_delivered_reinvoiced(self):
+        """
+        Tests that when we refund a re-invoiced expense, the Quantity Delivered on the Sale Order Line is updated
+        - (1) We create a Sale Order
+        - (2) We create a bill to be re-invoiced
+        - (3) We create a partial credit note
+        -> The sale order lines created in (2) should be updated during (3) with the correct delivered quantity.
+        """
+        # create the setup
+        product_1 = self.company_data['product_order_cost']
+        product_2 = self.env['product.product'].create({
+            'name': 'Great Product',
+            'standard_price': 50.0,
+            'list_price': 100.0,
+            'type': 'consu',
+            'uom_id': self.env.ref('uom.product_uom_unit').id,
+            'uom_po_id': self.env.ref('uom.product_uom_unit').id,
+            'invoice_policy': 'order',
+            'expense_policy': 'cost',
+        })
+        self._create_sol(self.sale_order, product_1)
+        self._create_sol(self.sale_order, product_2)
+        self.sale_order.action_confirm()
+
+        # create the bill to be re-invoiced
+        bill_form = Form(self.AccountMove.with_context(default_move_type='in_invoice'))
+        bill_form.partner_id = self.partner_b
+        with bill_form.line_ids.new() as line_form:
+            line_form.product_id = product_1
```

Comment:
Need to edit the `invoice_line_ids` to edit the product (see failing CI)

Reply:
Yep, but other issues related to analytic are at stake in 16.